### ISSUE TITLE
Working with BindJSON and error return

### ISSUE THREAD

**Issue by Depado** [gin-gonic/gin#815]
Hello,

First of all, thanks for your great project, I'm using it a on daily basis and it's just awesome.

## TL;DR

I have no clue on how to properly handle response format when validation fails on `BindJSON` and I want to send back the errors as JSON.

## Context

I'm currently having trouble working with `context.BindJSON`. I have two fields that are required when receiving JSON, let's call them `first` and `second`

```golang
type params struct {
	First  string `json:"first" binding:"required"`
	Second string `json:"second" binding:"required"`
}
```

Now when I receive the request, I handle it like this : 

```golang
func postFirstSecond(c *gin.Context) {
	var err error
	c.Header("Content-Type", "application/json; charset=utf-8")

	p := params{}
	if err = c.BindJSON(&p); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error":  "json decoding : " + err.Error(),
			"status": http.StatusBadRequest,
		})
		return
	}
}
```

So first of all, I don't understand why `c.JSON()` doesn't return the appropriate content-type, I found that when `BindJSON` fails, it writes the header to be plain text. Anyway my endpoint always returns JSON so I just set the content-type on top of the handler, problem solved (I guess).

When one field is missing, this endpoint returns something like that :

```json
{
    "error": "json decoding : Key: 'params.First' Error:Field validation for 'First' failed on the 'required' tag",
    "status": 400
}
```
That's fine. I mean it's not really what I expected but it works.

## Problem

When both fields are missing, this is what I get back :

```json
{
    "error": "json decoding : Key: 'params.First' Error:Field validation for 'First' failed on the 'required' tag\nKey: 'params.Second' Error:Field validation for 'Second' failed on the 'required' tag",
    "status": 400
}
```

## Question

Is there a simple way to handle those errors ? I'd like to get something more user-friendly, or at least a separate field for my errors. Something like :

```json
{
    "errors": [
        "Key: 'params.First' Error:Field validation for 'First' failed on the 'required' tag",
        "Key: 'params.Second' Error:Field validation for 'Second' failed on the 'required' tag"
     ],
    "status": 400
}
```

I'm aware that checking manually the fields should allow me to do that, but the fact is : I feel like the documentation and examples are pretty vague about error handling. 

Could someone help me with that ? How do you handle errors generated by `Bind` or `BindJSON` ?

**Comment by joeybloggs** [comment]
@Depado the error returned is actually of type [ValidationErrors](https://github.com/go-playground/validator/blob/v8/validator.go#L187) and you can cast it and then for or range over each FieldError in the map to get separate fields.

Also for those that may be using v9 of validator the definition is no longer a map https://github.com/go-playground/validator/blob/v9/errors.go#L37

**Comment by Depado** [comment]
Thanks @joeybloggs I'll give that a try 👍

**Comment by toefel18** [comment]
I also experience the error that the 'Content-Type' is text/plain instead of 'application/json' when I have a binding error and return a error struct using `c.JSON()`

**Comment by joeybloggs** [comment]
Yes you are correct @toefel18 not all errors that can be returned are `ValidationErrors` some are related to, as you pointed out, `Content-Type` and also JSON decoding errors.

so one would have to ensure the cast succeeds:

```go
err:= Bind...
ve, ok := err.(validator.ValidationErrors)
if !ok {
    // non validation error... handle appropriately.
}
```

**Comment by toefel18** [comment]
@joeybloggs that is indeed true, but the problem I am referring to is different
```go
if err := c.Bind(&login); err != nil {
     c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
}
```
Output is JSON, response Content-Type is "text/plain". I thought that is what @Depado meant. Should I file another bug request?

**Comment by joeybloggs** [comment]
I see @toefel18 although I am not a part of the gin team, I'd say yes it does seem like a separate issue.

P.S. may want to check there isn't one open already, for some reason it sounds familiar :)

**Comment by toefel18** [comment]
@joeybloggs, okay I will look into it later today :+1:

**Comment by Depado** [comment]
The Content-Type thing is a known issue.

**Comment by appleboy** [comment]
@Depado Try `c.AbortWithStatusJSON` function in gin `develop` branch. Please ref: #800

**Comment by toefel18** [comment]
Ah great, thanks @Depado @appleboy!

**Comment by appleboy** [comment]
@Depado Does `c.AbortWithStatusJSON` function resolve your problem abount `Content-Type`?

**Comment by Depado** [comment]
Looks like it does but the main point of this issue was about the parsing of the returned error of the `c.Bind` method.

**Comment by nazwa** [comment]
@depado you can use a middleware to check for any binding errors. There was a long thread about it somewhere here a while ago.

**Comment by AlexeyAtIgloo** [comment]
I've encountered similar issue, I have global after middleware that creates json response with errors when errors are present on the context and I also set header to json in that middleware, but because framework writes headers when handling bind errors, its impossible to overwrite them later...  My issue #840, I also have a workaround bind function i'm using for now, see my issue for it.

**Comment by mymtw** [comment]
I have the same issue
```
        var ucr messages.UsersCreateRequest
	if e := c.BindJSON(&ucr); e != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"errors": "someError/nHere/nAndYeatOneError",
		})
		return
	}
```
then sometimes I'm getting `text/plain`, but should `text/json`, sometimes I'm getting `text/json`

**Comment by easonlin404** [comment]
@mymtw f you want to get Content-Type application/json header when validating failed , you can use ShouldBindWith instead of Bind func that will not call AbortWithError

ref: https://github.com/gin-gonic/gin/issues/1041#issuecomment-319125864

**Comment by mymtw** [comment]
@easonlin404 thx, works

**Comment by appleboy** [comment]
Everybody can see the PR. https://github.com/gin-gonic/gin/pull/1047
